pkg/util/leaderlease/leaderlease.go

```diff
@@ -0,0 +1,269 @@
+package leaderlease
+
+import (
+	"fmt"
+	"time"
+
+	etcdclient "github.com/coreos/go-etcd/etcd"
+	"github.com/golang/glog"
+	storage "k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+// Leaser allows a caller to acquire a lease and be notified when it is lost.
+type Leaser interface {
+	// AcquireAndHold tries to acquire the lease and hold it until it expires, the lease is deleted,
+	// or we observe another party take the lease. The provided function will be invoked when the
+	// lease is acquired, and the provided channel will be closed when the lease is lost. If the
+	// function returns true, the lease will be released on exit. If the function returns false,
+	// the lease will be held.
+	AcquireAndHold(chan struct{})
+	// Release returns any active leases
+	Release()
+}
+
+// Etcd takes and holds a leader lease until it can no longer confirm it owns
+// the lease, then returns.
+type Etcd struct {
+	client *etcdclient.Client
+	key    string
+	value  string
+	ttl    uint64
+
+	// the fraction of the ttl to wait before trying to renew
+	waitFraction float32
+	// the interval to wait when an error occurs acquiring the lease
+	pauseInterval time.Duration
+	// the maximum retries to renew the lease
+	renewRetries int
+	// the shortest time between attempts to renew the lease
+	minimumRetryInterval time.Duration
+}
+
+// NewEtcd creates a Lease in etcd, storing value at key with expiration ttl
+// and continues to refresh it until the key is deleted, expires, or another
+// client takes it.
+func NewEtcd(client *etcdclient.Client, key, value string, ttl uint64) Leaser {
+	return &Etcd{
+		client: client,
+		key:    key,
+		value:  value,
+		ttl:    ttl,
+
+		waitFraction:         0.75,
+		pauseInterval:        time.Second,
+		renewRetries:         10,
+		minimumRetryInterval: 100 * time.Millisecond,
+	}
+}
+
+// AcquireAndHold implements an acquire and release of a lease.
+func (e *Etcd) AcquireAndHold(notify chan struct{}) {
+	for {
+		ok, ttl, index, err := e.tryAcquire()
+		if err != nil {
+			util.HandleError(err)
+			time.Sleep(e.pauseInterval)
+			continue
+		}
+		if !ok {
+			time.Sleep(e.pauseInterval)
+			continue
+		}
+
+		// notify
+		notify <- struct{}{}
+		defer close(notify)
+
+		// hold the lease
+		if err := e.tryHold(ttl, index); err != nil {
+			util.HandleError(err)
+		}
+		break
+	}
+}
+
+// tryAcquire tries to create the lease key in etcd, or if it already exists
+// and belongs to another user, to wait until the lease expires or is deleted.
+func (e *Etcd) tryAcquire() (ok bool, ttl uint64, index uint64, err error) {
```

Comment:
I had to read to see what `index` was used for to decide if `+1` makes sense.  `startingWatchIndex`?